File: selfjavagame/src/trrne/Lottery.java
package trrne;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

public class Lottery<TSubject> {
    public static int binarySearchTree(List<Double> $weights) throws Exception {
        if ($weights.isEmpty()) {
            throw new Exception();
        }

        List<Double> totals = new ArrayList<>($weights.size());
        double total = 0.f;
        for (int i = 0; i < $weights.size(); i++) {
            total += $weights.get(i);
            totals.add(total);
        }

        // double rnd = Mathf.randuniform(0f, total);
        double rnd = new Random().nextDouble(0f, total);
        int bottom = 0, top = totals.size() - 1;
        while (bottom < top) {
            int middle = (bottom + top) / 2;
            if (rnd > totals.get(middle)) {
                bottom = middle + 1;
            } else {
                double point = middle > 0 ? totals.get(middle - 1) : 0;
                if (rnd >= point) {
                    return middle;
                }
                top = middle;
            }
        }
        return top;
    }

    public static int bst(Double[] $weights) throws Exception {
        return binarySearchTree(Arrays.asList($weights));
    }
}

File: selfcppgame2/lottery.cpp
#include "lottery.h"

using namespace trrne;

int lottery::bst(float *weights_arr, const size_t length) {
	if (length <= 0) {
		throw exception("karappo yanke");
	}

	vector<float> totals;
	float total = 0.f;
	for (int i = 0; i < length; ++i) {
		total += weights_arr[i];
		totals.push_back(total);
	}

	float r = rnd::randfloat(0, total);
	int top = int(length) - 1, bottom = 0;
	while (bottom < top) {
		int center = (bottom + top) / 2;
		if (r > totals[center]) {
			bottom = center + 1;
		} else {
			float p = center > 0 ? totals[size_t(center - 1)] : 0.f;
			if (r >= p) {
				return center;
			}
			top = center;
		}
	}
	return top;
}
